sqlanalytics/resource_query.go

```diff
@@ -0,0 +1,495 @@
+					Multiple: newQueryParameterAllowMultiple(apv.Multi),
+				}
+				if p.Query.Multiple != nil {
+					p.Query.Values = apv.Values
+				} else {
+					p.Query.Value = apv.Values[0]
+				}
+			case *api.QueryParameterDate:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.Date = &QueryParameterDate{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTime:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTime = &QueryParameterDateTime{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeSec:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeSec = &QueryParameterDateTimeSec{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateRange = &QueryParameterDateRange{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeRange = &QueryParameterDateTimeRange{
+					Value: apv.Value,
+				}
+			case *api.QueryParameterDateTimeSecRange:
+				p.Name = apv.Name
+				p.Title = apv.Title
+				p.DateTimeSecRange = &QueryParameterDateTimeSecRange{
+					Value: apv.Value,
+				}
+			default:
+				log.Fatalf("Don't know what to do for type: %#v", reflect.TypeOf(apv).String())
+			}
+
+			q.Parameter = append(q.Parameter, p)
+		}
+	}
+
+	// Transform to ResourceData.
+	if err := common.StructToData(*q, schema, data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewQueryAPI ...
+func NewQueryAPI(ctx context.Context, m interface{}) QueryAPI {
+	return QueryAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// QueryAPI ...
+type QueryAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+func (a QueryAPI) buildPath(path ...string) string {
+	out := "/preview/sql/queries"
+	if len(path) == 1 {
+		out = out + "/" + strings.Join(path, "/")
+	}
+	return out
+}
+
+// Create ...
+func (a QueryAPI) Create(q *api.Query) (*api.Query, error) {
+	var qp api.Query
+	err := a.client.Post(a.context, a.buildPath(), q, &qp)
+	if err != nil {
+		return nil, err
+	}
+
+	// New queries are created with a table visualization by default.
+	// We don't manage that visualization here, so immediately remove it.
+	if len(qp.Visualizations) > 0 {
+		for _, rv := range qp.Visualizations {
+			var v api.Visualization
+			err = json.Unmarshal(rv, &v)
+			if err != nil {
+				return nil, err
+			}
+			// TODO
+			// err = a.DeleteVisualization(&v)
+			// if err != nil {
+			// 	return nil, err
+			// }
+		}
+		qp.Visualizations = []json.RawMessage{}
+	}
+
+	return &qp, err
+}
+
+// Read ...
+func (a QueryAPI) Read(q *api.Query) (*api.Query, error) {
+	var qp api.Query
+	err := a.client.Get(a.context, a.buildPath(q.ID), nil, &qp)
+	if err != nil {
+		return nil, err
+	}
+
+	return &qp, nil
+}
+
+// Update ...
+func (a QueryAPI) Update(q *api.Query) (*api.Query, error) {
+	var qp api.Query
+	err := a.client.Post(a.context, a.buildPath(q.ID), q, &qp)
```

Comment:
i just wonder if we could have `q, &q`? e.g. if we cannot - then what is the reason? struct is the same, entity is the same, parameter accepts the pointer, so there's no immutability, that asks for a new instance in return 🤷

Reply:
I used a new instance here to ensure you don't pass along input parameters that got removed as output.